spec/api-web-contents-spec.js

```diff
@@ -542,4 +542,70 @@ describe('webContents module', function () {
       })
     })
   })
+
+  describe('destroy()', () => {
+    let server
+
+    before(function (done) {
+      server = http.createServer((request, response) => {
+        switch (request.url) {
+          case '/404':
+            response.statusCode = '404'
+            response.end()
+            break
+          case '/301':
+            response.statusCode = '301'
+            response.setHeader('Location', '/200')
+            response.end()
+            break
+          case '/200':
+            response.statusCode = '200'
+            response.end('hello')
+            break
+          default:
+            done('unsupported endpoint')
+        }
+      }).listen(0, '127.0.0.1', () => {
+        server.url = 'http://127.0.0.1:' + server.address().port
+        done()
+      })
+    })
+
+    after(function () {
+      server.close()
+      server = null
+    })
+
+    it('should not crash when invoked synchronously inside navigation observer', (done) => {
+      const events = [
+        { name: 'did-start-loading', url: `${server.url}/200` },
+        { name: 'did-get-redirect-request', url: `${server.url}/301` },
+        { name: 'did-get-response-details', url: `${server.url}/200` },
+        { name: 'dom-ready', url: `${server.url}/200` },
+        { name: 'did-stop-loading', url: `${server.url}/200` },
+        { name: 'did-finish-load', url: `${server.url}/200` },
+        // FIXME: Multiple Emit calls inside an observer assume that object
+        // will be alive till end of the observer. Synchronous `destroy` api
+        // violates this contract and crashes.
+        // { name: 'did-frame-finish-load', url: `${server.url}/200` },
```

Comment:
@zcbenz currently we have in `api_web_contents.cc` the following:

```
void WebContents::DidFinishLoad(content::RenderFrameHost* render_frame_host,
                                const GURL& validated_url) {
  bool is_main_frame = !render_frame_host->GetParent();
  Emit("did-frame-finish-load", is_main_frame);

  if (is_main_frame)
    Emit("did-finish-load");
}
```

calling `webContents.destroy` inside `did-frame-finish-load` will lead to a crash when `did-finish-load` is invoked with
```
[2853:0404/235125.714735:FATAL:wrappable.cc(24)] Check failed: !wrapper_.IsEmpty(). 
#0 0x7fe05839616e base::debug::StackTrace::StackTrace()
#1 0x7fe0583b2eab logging::LogMessage::~LogMessage()
#2 0x000000b8d723 mate::WrappableBase::GetWrapper()
#3 0x000000c782e5 mate::EventEmitter<>::GetWrapper()
#4 0x000000c7b1f9 _ZN4mate12EventEmitterIN4atom3api11WebContentsEE14EmitWithSenderIJEEEbRKN4base16BasicStringPieceISsEEPN7content11WebContentsEPN3IPC7MessageEDpRKT_
#5 0x000000c71bf4 _ZN4mate12EventEmitterIN4atom3api11WebContentsEE4EmitIJEEEbRKN4base16BasicStringPieceISsEEDpRKT_
#6 0x000000c6761a atom::api::WebContents::DidFinishLoad()
#7 0x7fe05c397999 content::WebContentsImpl::OnDidFinishLoad()
#8 0x7fe05c39777e <unknown>
#9 0x7fe05c3963fc content::WebContentsImpl::OnMessageReceived()
#10 0x7fe05c121888 content::RenderFrameHostImpl::OnMessageReceived()
#11 0x7fe057edeac5 IPC::ChannelProxy::Context::OnDispatchMessage()
#12 0x7fe058396cee base::debug::TaskAnnotator::RunTask()
#13 0x7fe0583bba1c base::MessageLoop::RunTask()
#14 0x7fe0583bbd38 base::MessageLoop::DeferOrRunPendingTask()
#15 0x7fe0583bc0fb base::MessageLoop::DoWork()
#16 0x7fe0583be4aa <unknown>
#17 0x7fe0499835a7 g_main_context_dispatch
#18 0x7fe049983810 <unknown>
#19 0x7fe0499838bc g_main_context_iteration
#20 0x7fe0583be316 base::MessagePumpGlib::Run()
#21 0x7fe0583bb7b7 base::MessageLoop::RunHandler()
#22 0x7fe0583e52b0 base::RunLoop::Run()
#23 0x7fe05c03bab8 content::BrowserMainLoop::MainMessageLoopRun()
#24 0x7fe05c03b905 content::BrowserMainLoop::RunMainMessageLoopParts()
#25 0x7fe05c03e94d <unknown>
#26 0x7fe05c03743e content::BrowserMain()
#27 0x7fe05c6db094 <unknown>
#28 0x7fe05c6d9ba0 content::ContentMain()
#29 0x000000aa569a main
#30 0x7fe047ae9511 __libc_start_main
#31 0x000000aa54e5 <unknown>
```

I can think of two ways to fix this:
* Make `destroy` async by default https://github.com/deepak1556/native-mate/commit/c98d5d6ae420831e7db9ab1fb81ffdc5eb6728cd , will have to rewire cases inside the code base to follow this behavior, especially webview since we need to make sure they are destroyed before any host webContents.

* Return empty wrapper when its destroyed and check if object is available before emitting events.

What can be the best approach to fix this scenario ?